File: internal/chat/tool_policy_matcher.go
package chat

import (
	"strings"

	mcplib "assistente/internal/mcp"
)

// ToolPolicyTarget reúne os atributos usados pela gramática de seletores.
// Package deve vir de tools.CatalogMetadata para builtins e ficar vazio para
// tools MCP, cuja seleção é feita pelo namespace canônico.
type ToolPolicyTarget struct {
	Name    string
	Package string
	OptIn   bool
}

type ToolPolicySelectorKind int

const (
	ToolPolicySelectorLiteral ToolPolicySelectorKind = iota
	ToolPolicySelectorAllNative
	ToolPolicySelectorAllMCP
	ToolPolicySelectorAllPackages
	ToolPolicySelectorMCPServer
	ToolPolicySelectorPackage
)

// ToolPolicySelector é a representação compilada e reutilizável da gramática
// de tool_policy. A issue #630 pode usá-la para expandir load sem reimplementar
// aliases, namespaces ou precedência.
type ToolPolicySelector struct {
	Canonical string
	Kind      ToolPolicySelectorKind
	Value     string
}

// ParseToolPolicySelector aceita literais e os wildcards documentados:
// *, mcp/*, mcp/<slug>/*, mcp:<slug>/*, package/*,
// package/<pacote>/* e <pacote>/*. Também aceita a forma canônica interna
// mcp_<slug>__* (e mcp_*__*).
func ParseToolPolicySelector(raw string) (ToolPolicySelector, bool) {
	value := strings.TrimSpace(raw)
	if value == "" {
		return ToolPolicySelector{}, false
	}
	if value == "*" {
		return ToolPolicySelector{Canonical: "*", Kind: ToolPolicySelectorAllNative}, true
	}
	if value == "mcp/*" || value == "mcp_*__*" {
		return ToolPolicySelector{Canonical: "mcp/*", Kind: ToolPolicySelectorAllMCP}, true
	}
	if value == "package/*" {
		return ToolPolicySelector{Canonical: "package/*", Kind: ToolPolicySelectorAllPackages}, true
	}

	if slug, ok := parseDelimitedSelector(value, "mcp/", "/*"); ok {
		return mcpServerSelector(slug)
	}
	if slug, ok := parseDelimitedSelector(value, "mcp:", "/*"); ok {
		return mcpServerSelector(slug)
	}
	// Compatibilidade com a grafia mcp__<slug>__* usada em documentação
	// histórica, embora o nome executável atual seja mcp_<slug>__<tool>.
	if slug, ok := parseDelimitedSelector(value, "mcp__", "__*"); ok {
		return mcpServerSelector(slug)
	}
	if slug, ok := parseDelimitedSelector(value, "mcp_", "__*"); ok {
		return mcpServerSelector(slug)
	}
	if pkg, ok := parseDelimitedSelector(value, "package/", "/*"); ok {
		return packageSelector(pkg)
	}
	if pkg, ok := parseDelimitedSelector(value, "", "/*"); ok && !strings.Contains(pkg, "/") {
		return packageSelector(pkg)
	}

	return ToolPolicySelector{
		Canonical: value,
		Kind:      ToolPolicySelectorLiteral,
		Value:     value,
	}, true
}

func parseDelimitedSelector(value, prefix, suffix string) (string, bool) {
	if !strings.HasPrefix(value, prefix) || !strings.HasSuffix(value, suffix) {
		return "", false
	}
	middle := strings.TrimSpace(strings.TrimSuffix(strings.TrimPrefix(value, prefix), suffix))
	return middle, middle != "" && middle != "*"
}

func mcpServerSelector(slug string) (ToolPolicySelector, bool) {
	slug = strings.TrimSpace(slug)
	if slug == "" || strings.ContainsAny(slug, "/*:") {
		return ToolPolicySelector{}, false
	}
	return ToolPolicySelector{
		Canonical: "mcp/" + slug + "/*",
		Kind:      ToolPolicySelectorMCPServer,
		Value:     slug,
	}, true
}

func packageSelector(pkg string) (ToolPolicySelector, bool) {
	pkg = strings.TrimSpace(pkg)
	if pkg == "" || strings.ContainsAny(pkg, "/*:") {
		return ToolPolicySelector{}, false
	}
	return ToolPolicySelector{
		Canonical: "package/" + pkg + "/*",
		Kind:      ToolPolicySelectorPackage,
		Value:     pkg,
	}, true
}

func (s ToolPolicySelector) Matches(target ToolPolicyTarget) bool {
	switch s.Kind {
	case ToolPolicySelectorLiteral:
		return target.Name == s.Value
	case ToolPolicySelectorAllNative:
		_, _, isMCP := mcplib.ParseToolName(target.Name)
		return !isMCP
	case ToolPolicySelectorAllMCP:
		_, _, isMCP := mcplib.ParseToolName(target.Name)
		return isMCP
	case ToolPolicySelectorAllPackages:
		return strings.TrimSpace(target.Package) != ""
	case ToolPolicySelectorMCPServer:
		slug, _, isMCP := mcplib.ParseToolName(target.Name)
		return isMCP && slug == s.Value
	case ToolPolicySelectorPackage:
		return target.Package == s.Value
	default:
		return false
	}
}

// Specificity ordena default < wildcard geral < wildcard específico < literal.
func (s ToolPolicySelector) Specificity() int {
	switch s.Kind {
	case ToolPolicySelectorLiteral:
		return 3
	case ToolPolicySelectorMCPServer, ToolPolicySelectorPackage:
		return 2
	default:
		return 1
	}
}

type toolPolicyRule struct {
	selector ToolPolicySelector
	state    ToolPolicyState
}

// ToolPolicyMatch descreve tanto o estado resolvido quanto sua origem.
type ToolPolicyMatch struct {
	State       ToolPolicyState
	Selector    string
	Specificity int
	Explicit    bool
	Literal     bool
	DeniedOptIn bool
}

// ToolPolicyMatcher compila uma política uma vez e a aplica de modo lazy a
// tools que podem entrar no registry depois da criação da política.
type ToolPolicyMatcher struct {
	rules        []toolPolicyRule
	defaultState ToolPolicyState
}

func NewToolPolicyMatcher(configured map[string]string, defaultState string) ToolPolicyMatcher {
	matcher := ToolPolicyMatcher{defaultState: normalizeToolPolicyDefault(defaultState)}
	normalized := make(map[string]toolPolicyRule, len(configured))
	for raw, rawState := range configured {
		selector, ok := ParseToolPolicySelector(raw)
		if !ok {
			continue
		}
		rule := toolPolicyRule{selector: selector, state: normalizeToolPolicyState(rawState)}
		if existing, found := normalized[selector.Canonical]; found &&
			toolPolicyStateRank(existing.state) <= toolPolicyStateRank(rule.state) {
			continue
		}
		normalized[selector.Canonical] = rule
	}
	for _, rule := range normalized {
		matcher.rules = append(matcher.rules, rule)
	}
	return matcher
}

func (m ToolPolicyMatcher) Resolve(target ToolPolicyTarget) ToolPolicyMatch {
	best := ToolPolicyMatch{State: m.defaultState}
	for _, rule := range m.rules {
		if !rule.selector.Matches(target) {
			continue
		}
		specificity := rule.selector.Specificity()
		if best.Explicit && specificity < best.Specificity {
			continue
		}
		if best.Explicit && specificity == best.Specificity &&
			toolPolicyStateRank(best.State) <= toolPolicyStateRank(rule.state) {
			continue
		}
		best = ToolPolicyMatch{
			State:       rule.state,
			Selector:    rule.selector.Canonical,
			Specificity: specificity,
			Explicit:    true,
			Literal:     rule.selector.Kind == ToolPolicySelectorLiteral,
		}
	}
	// Opt-in exige autorização literal; default e wildcard permissivos jamais
	// elevam a capability.
	if target.OptIn && best.State != ToolPolicyDisabled && !best.Literal {
		best.State = ToolPolicyDisabled
		best.DeniedOptIn = true
	}
	return best
}

func (m ToolPolicyMatcher) HasOnDemandRule() bool {
	for _, rule := range m.rules {
		if rule.state == ToolPolicyOnDemand {
			return true
		}
	}
	return false
}

File: frontend/src/utils/toolPolicyMatcher.ts
import { parseToolSource } from './toolSource';

export type ToolPolicyState = 'disabled' | 'on_demand' | 'preloaded';

export interface ToolPolicyTarget {
  name: string;
  package?: string;
  optIn?: boolean;
}

type SelectorKind = 'literal' | 'native-all' | 'mcp-all' | 'package-all' | 'mcp-server' | 'package';

interface Selector {
  canonical: string;
  kind: SelectorKind;
  value?: string;
  specificity: number;
}

export interface ToolPolicyMatch {
  state: ToolPolicyState;
  selector?: string;
  specificity: number;
  explicit: boolean;
  literal: boolean;
  deniedOptIn?: boolean;
}

function delimited(value: string, prefix: string, suffix: string): string | null {
  if (!value.startsWith(prefix) || !value.endsWith(suffix)) return null;
  const middle = value.slice(prefix.length, value.length - suffix.length).trim();
  return middle !== '' && middle !== '*' ? middle : null;
}

function scopedSelector(kind: 'mcp-server' | 'package', value: string): Selector | null {
  if (value === '' || /[/*:]/.test(value)) return null;
  return {
    canonical: kind === 'mcp-server' ? `mcp/${value}/*` : `package/${value}/*`,
    kind,
    value,
    specificity: 2,
  };
}

export function parseToolPolicySelector(raw: string): Selector | null {
  const value = raw.trim();
  if (value === '') return null;
  if (value === '*') return { canonical: '*', kind: 'native-all', specificity: 1 };
  if (value === 'mcp/*' || value === 'mcp_*__*') {
    return { canonical: 'mcp/*', kind: 'mcp-all', specificity: 1 };
  }
  if (value === 'package/*') {
    return { canonical: 'package/*', kind: 'package-all', specificity: 1 };
  }

  const mcpSlash = delimited(value, 'mcp/', '/*');
  if (mcpSlash) return scopedSelector('mcp-server', mcpSlash);
  const mcpColon = delimited(value, 'mcp:', '/*');
  if (mcpColon) return scopedSelector('mcp-server', mcpColon);
  const historicalMcp = delimited(value, 'mcp__', '__*');
  if (historicalMcp) return scopedSelector('mcp-server', historicalMcp);
  const canonicalMcp = delimited(value, 'mcp_', '__*');
  if (canonicalMcp) return scopedSelector('mcp-server', canonicalMcp);
  const packageSlash = delimited(value, 'package/', '/*');
  if (packageSlash) return scopedSelector('package', packageSlash);
  const shortPackage = delimited(value, '', '/*');
  if (shortPackage && !shortPackage.includes('/')) return scopedSelector('package', shortPackage);

  return { canonical: value, kind: 'literal', value, specificity: 3 };
}

function selectorMatches(selector: Selector, target: ToolPolicyTarget): boolean {
  const source = parseToolSource(target.name);
  switch (selector.kind) {
    case 'literal': return target.name === selector.value;
    case 'native-all': return source.type !== 'mcp';
    case 'mcp-all': return source.type === 'mcp';
    case 'package-all': return (target.package?.trim() ?? '') !== '';
    case 'mcp-server': return source.type === 'mcp' && source.serverSlug === selector.value;
    case 'package': return target.package === selector.value;
  }
}

export function normalizeToolPolicyState(state: string): ToolPolicyState {
  const normalized = state.trim();
  if (normalized === 'on_demand' || normalized === 'preloaded') return normalized;
  return 'disabled';
}

function stateRank(state: ToolPolicyState): number {
  if (state === 'disabled') return 0;
  if (state === 'on_demand') return 1;
  return 2;
}

export function normalizeToolPolicyMap(policy: Record<string, string> | null): Record<string, string> {
  const normalized: Record<string, string> = {};
  for (const [raw, state] of Object.entries(policy ?? {})) {
    const selector = parseToolPolicySelector(raw);
    if (!selector) continue;
    const existing = normalized[selector.canonical];
    if (existing != null
      && stateRank(normalizeToolPolicyState(existing)) <= stateRank(normalizeToolPolicyState(state))) {
      continue;
    }
    normalized[selector.canonical] = state;
  }
  return normalized;
}

export function resolveToolPolicy(
  policy: Record<string, string>,
  defaultState: string | null | undefined,
  target: ToolPolicyTarget,
): ToolPolicyMatch {
  const fallback = defaultState?.trim() === 'on_demand' ? 'on_demand' : 'disabled';
  let best: ToolPolicyMatch = {
    state: fallback,
    specificity: 0,
    explicit: false,
    literal: false,
  };
  for (const [raw, rawState] of Object.entries(policy)) {
    const selector = parseToolPolicySelector(raw);
    if (!selector || !selectorMatches(selector, target)) continue;
    const state = normalizeToolPolicyState(rawState);
    if (best.explicit && selector.specificity < best.specificity) continue;
    if (best.explicit && selector.specificity === best.specificity
      && stateRank(best.state) <= stateRank(state)) continue;
    best = {
      state,
      selector: selector.canonical,
      specificity: selector.specificity,
      explicit: true,
      literal: selector.kind === 'literal',
    };
  }
  if (target.optIn && best.state !== 'disabled' && !best.literal) {
    return { ...best, state: 'disabled', deniedOptIn: true };
  }
  return best;
}
